rootfs/apply.go

```diff
@@ -47,76 +48,123 @@ type Layer struct {
 // Layers are applied in order they are given, making the first layer the
 // bottom-most layer in the layer chain.
 func ApplyLayers(ctx context.Context, layers []Layer, sn snapshots.Snapshotter, a diff.Applier) (digest.Digest, error) {
-	var chain []digest.Digest
-	for _, layer := range layers {
-		if _, err := ApplyLayer(ctx, layer, chain, sn, a); err != nil {
-			// TODO: possibly wait and retry if extraction of same chain id was in progress
-			return "", err
+	chain := make([]digest.Digest, len(layers))
+	for i, layer := range layers {
+		chain[i] = layer.Diff.Digest
+	}
+	chainID := identity.ChainID(chain)
+
+	var tries int
+	for {
+		// Just stat top layer, remaining layers will have their existence checked
+		// on prepare. Calling prepare on upper layers first guarantees that upper
+		// layers are not removed while calling stat on lower layers
+		_, err := sn.Stat(ctx, chainID.String())
+		if err != nil {
+			tries++
+			if tries < 3 && errdefs.IsNotFound(err) {
+				if err := applyLayers(ctx, layers, chain, sn, a); err != nil {
+					if errdefs.IsUnavailable(err) {
+						time.Sleep(time.Second * time.Duration(tries))
```

Comment:
Do we have to wait at a min 1 sec?  Seems a little long but not sure.

Reply:
I can simplify this and get rid of the unavailability checks by moving the key generation into the loop. Likely this scenario cannot be reconciled and is reliant on cleanup anyway.